target/linux/ramips/dts/mt7621_belkin_rt1800.dts

```diff
@@ -0,0 +1,180 @@
+// SPDX-License-Identifier: GPL-2.0-or-later OR MIT
+
+#include "mt7621.dtsi"
+
+#include <dt-bindings/gpio/gpio.h>
+#include <dt-bindings/input/input.h>
+#include <dt-bindings/leds/common.h>
+
+/ {
+	compatible = "belkin,rt1800", "mediatek,mt7621-soc";
+	model = "Belkin RT1800";
+
+	aliases {
+		led-boot = &led_power;
+		led-failsafe = &led_power;
+		led-running = &led_power;
+		led-upgrade = &led_power;
+	};
+
+	chosen {
+		bootargs = "console=ttyS0,115200";
+	};
+
+	keys {
+		compatible = "gpio-keys";
+
+		reset {
+			label = "reset";
+			gpios = <&gpio 6 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
+
+		wps {
+			label = "wps";
+			gpios = <&gpio 18 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_WPS_BUTTON>;
+		};
+	};
+
+	leds {
+		compatible = "gpio-leds";
+
+		led-wps {
+			color = <LED_COLOR_ID_AMBER>;
+			function = LED_FUNCTION_WPS;
+			gpios = <&gpio 13 GPIO_ACTIVE_LOW>;
+		};
+
+		led_power: led-power {
+			label = "white:power";
+			gpios = <&gpio 14 GPIO_ACTIVE_LOW>;
+			default-state = "on";
+		};
+
+		led-wan {
+			color = <LED_COLOR_ID_AMBER>;
+			function = LED_FUNCTION_WAN;
+			gpios = <&gpio 15 GPIO_ACTIVE_LOW>;
+		};
+
+		led-wan2 {
+			color = <LED_COLOR_ID_WHITE>;
+			function = LED_FUNCTION_WAN;
+			gpios = <&switch0 0 GPIO_ACTIVE_LOW>;
+		};
+	};
+};
+
+&nand {
+	status = "okay";
+
+	mediatek,bbt;
+	mediatek,bmt-remap-range = <0x0180000 0x0580000>;
```

Comment:
It seems that all partitions except `ubi` should be remapped. b86452f8414d921058d800922849691a0fcfe2a6

Reply:
This part I don't get.

I asked @nbd168 on IRC and he recommended only doing it for the kernel for BMTv1. I forgot the details.

Reply:
Looking at the code, I'm wondering if this should be mediatek,nmbm instead...

Reply:
Well, I see it. e2aa04d1e50104456902aea685e46c754b07ebf4
These bad block management algorithms are used to help u-boot locate and load the kernel, so both need to use the same algorithm. At least the GPL code shows that it doesn't support nmbm.

Reply:
dmesg says this:

```
[    1.363654] nand: device found, Manufacturer ID: 0xef, Chip ID: 0xf1
[    1.370079] nand: Winbond W29N01HV
[    1.373486] nand: 128 MiB, SLC, erase size: 128 KiB, page size: 2048, OOB size: 64
[    1.381047] mt7621-nand 1e003000.nand: ECC strength adjusted to 4 bits
[    1.388456] read_bbt: signature mismatch in block 1023
[    1.393603] 00000000: ff ff ff ff ff ff ff ff                          ........
[    1.401535] read_bbt: signature mismatch in block 1022
[    1.406683] 00000000: ff ff ff ff ff ff ff ff                          ........
[    1.414596] read_bbt: signature mismatch in block 1021
[    1.419719] 00000000: ff ff ff ff ff ff ff ff                          ........
[    1.427643] read_bbt: signature mismatch in block 1020
[    1.432761] 00000000: ff ff ff ff ff ff ff ff                          ........
[    1.440677] read_bbt: signature mismatch in block 1019
[    1.445794] 00000000: ff ff ff ff ff ff ff ff                          ........

```

Reply:
I can confirm that bbt is used in the E7350's u-boot GPL code:
```
CONFIG_COMPAT_NAND_BBT=y
CONFIG_COMPAT_NAND_BBT_WB=y
```

Maybe need to add `nand-ecc-mode = "hw";`? Or just drop bbt? Not sure.

<details>
<summary>GPL</summary>

```
&nand {
	status = "okay";

	flash@0 {
		reg = <0>;
		nand-ecc-mode = "hw";
		partitions {
			compatible = "fixed-partitions";
			#address-cells = <1>;
			#size-cells = <1>;

			partition@0 {
				label = "Bootloader";
				reg = <0x00000 0x80000>;
				read-only;
			};

			partition@80000 {
				label = "Config";
				reg = <0x80000 0x80000>;
			};

			partition@100000 {
				label = "Factory";
				reg = <0x100000 0x80000>;
			};

			partition@180000 {
				label = "firmware";
				reg = <0x180000 0x3000000>;
			};
			partition@3180000 {
				label = "firmware_2";
				reg = <0x3180000 0x3000000>;
			};
			partition@6180000 {
				label = "CBTinfo";
				reg = <0x6180000 0x80000>;
			};
		};
	};
};
```
</details>

Reply:
the mtknand driver has a check for hw. It would fail earlier if it was not.

Reply:
maybe I get it, `CBTinfo` pattition = custom bad block table info? Seems Linksys redefines bbt offset? It doesn't make sense.

Reply:
It's checking for a block with mtknand as a header. string /dev/mtdX is not finding it.

Reply:
```
root@OpenWrt:~# hexdump /dev/mtd7ro 
0000000 9fe8 2c80 2389 9fe8 2c80 2489 9fe8 2c80
0000010 2589 9fe8 2c80 2689 ffff ffff ffff ffff
0000020 ffff ffff ffff ffff ffff ffff ffff ffff
*
0000100 3034 3153 4d30 4132 3041 3731 3233 0000
0000110 0000 0000 0000 0000 0000 0000 0000 0000
0000120 ffff ffff ffff ffff ffff ffff ffff ffff
*
0000280 7075 6d78 7875 6d70 7074 0000 0000 0000
0000290 0000 0000 0000 0000 0000 0000 0000 0000
*
0000300 ffff ffff ffff ffff ffff ffff ffff ffff
*
0000340 7075 6d78 7875 6d70 7074 0000 0000 0000
0000350 0000 0000 0000 0000 0000 0000 0000 0000
*
00003c0 ffff ffff ffff ffff ffff ffff ffff ffff
*
0000400 666f 0066 0000 0000 ffff ffff ffff ffff
0000410 5355 0000 0000 0000 3033 3032 3735 3330
0000420 0000 0000 0000 0000 5452 3831 3030 0000
0000430 0000 0000 0000 0000 ffff ffff ffff ffff
0000440 ffff ffff ffff ffff ffff ffff ffff ffff
*
0080000

```

Reply:
mtknand signature is located at oob and needs to be queried using u-boot's nand dump command. mtd HAL will hide oob automatically.

Reply:
so I need to open the router again...

Reply:
Or just drop bbt. Removing bbt shouldn't have much impact. It says to be compatible with old u-boot, we didn't use any bad block management on mt7621 before.
https://github.com/DragonBluep/uboot-mt7621/blob/55d826540bbc77e514eba320c90b382cc41e3816/board/ralink/mt7621_nand_rfb/Kconfig#L16-L33

Reply:
In bc7d36ba3a43bc3bc4eeab6ea127032aba3e1f4e Raisecom MSG1500's full nand backup file I can find this `mtknand` signature in block 1023 (0x7fe0000). The last line is oob.
```
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff ff
ff 6d 74 6b 6e 61 6e 64 dc 7e 7e e8 07 ec f8 ff
```

Reply:
Anyway I removed it.